connexion/middleware/main.py

```diff
@@ -1,19 +1,86 @@
+import dataclasses
+import logging
 import pathlib
 import typing as t
+from dataclasses import dataclass, field
 
 from starlette.types import ASGIApp, Receive, Scope, Send
 
-from connexion.middleware.abstract import AppMiddleware
+from connexion import utils
+from connexion.handlers import ResolverErrorHandler
+from connexion.middleware.abstract import SpecMiddleware
 from connexion.middleware.context import ContextMiddleware
 from connexion.middleware.exceptions import ExceptionMiddleware
 from connexion.middleware.request_validation import RequestValidationMiddleware
 from connexion.middleware.response_validation import ResponseValidationMiddleware
 from connexion.middleware.routing import RoutingMiddleware
 from connexion.middleware.security import SecurityMiddleware
 from connexion.middleware.swagger_ui import SwaggerUIMiddleware
+from connexion.resolver import Resolver
+from connexion.uri_parsing import AbstractURIParser
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class _Options:
+    """
+    Connexion provides a lot of parameters for the user to configure the app / middleware of
+    application.
+
+    This class provides a central place to parse these parameters a mechanism to update them.
+    Application level arguments can be provided when instantiating the application / middleware,
+    after which they can be overwritten on an API level.
+
+    The defaults should only be set in this class, and set to None in the signature of user facing
+    methods. This is necessary for this class to be able to differentiate between missing and
+    falsy arguments.
+    """
+
+    arguments: t.Optional[dict] = None
+    auth_all_paths: t.Optional[bool] = False
+    pythonic_params: t.Optional[bool] = False
+    resolver: t.Optional[t.Union[Resolver, t.Callable]] = None
+    resolver_error: t.Optional[int] = None
+    resolver_error_handler: t.Optional[t.Callable] = field(init=False)
+    strict_validation: t.Optional[bool] = False
+    swagger_ui_options: t.Optional[dict] = None
+    uri_parser_class: t.Optional[AbstractURIParser] = None
+    validate_responses: t.Optional[bool] = False
+    validator_map: t.Optional[dict] = None
+
+    def __post_init__(self):
+        self.resolver = (
+            Resolver(self.resolver) if callable(self.resolver) else self.resolver
+        )
+        self.resolver_error_handler = self._resolver_error_handler_factory()
+
+    def _resolver_error_handler_factory(
+        self,
+    ) -> t.Optional[t.Callable[[], ResolverErrorHandler]]:
+        """Returns a factory to create a ResolverErrorHandler."""
+        if self.resolver_error is not None:
+
+            def resolver_error_handler(*args, **kwargs) -> ResolverErrorHandler:
+                return ResolverErrorHandler(self.resolver_error, *args, **kwargs)
+
+            return resolver_error_handler
+        return None
+
+    def replace(self, **changes) -> "_Options":
```

Comment:
Not ideal that this is a separate function that needs to be called (if I understand correctly) - is there a way to make it part of the initialization of the class?

Reply:
This method is called after the initialization to update the instance with new options.

The instance is created in the Middleware `__init__` method and updated in the `add_api` method. This way, we can set default options for the middleware and override them on an API level.

Reply:
Got it now, the split is intentional :) I thought it was always something like
```python
options = _Options(...)
options.replace(...)
```